src/data/dataset.c

```
@@ -0,0 +1,138 @@
+}
+
+void free_tensor_dataset(TensorDataset *dataset) {
+  if (dataset) {
+    free(dataset->tensors);
+    free(dataset);
+  }
+}
+
+DataLoader *create_dataloader(TensorDataset *dataset, u64 batch_size, bool shuffle, bool drop_last, DEVICE device) {
+  DataLoader *loader = malloc(sizeof(DataLoader));
+  loader->dataset = dataset;
```

Cppcheck (2.20.0)

[warning] If memory allocation fails, then there is a possible null pointer dereference

(nullPointerOutOfMemory)